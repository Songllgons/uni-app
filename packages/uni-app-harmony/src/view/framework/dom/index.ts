import {
  ACTION_TYPE_ADD_EVENT,
  ACTION_TYPE_ADD_WXS_EVENT,
  ACTION_TYPE_CREATE,
  ACTION_TYPE_INSERT,
  ACTION_TYPE_PAGE_CREATE,
  ACTION_TYPE_PAGE_CREATED,
  ACTION_TYPE_PAGE_SCROLL,
  ACTION_TYPE_REMOVE,
  ACTION_TYPE_REMOVE_ATTRIBUTE,
  ACTION_TYPE_REMOVE_EVENT,
  ACTION_TYPE_SET_ATTRIBUTE,
  ACTION_TYPE_SET_TEXT,
  type PageAction,
  type PageCreateAction,
  formatLog,
} from '@dcloudio/uni-shared'
import type { UniNodeJSONMinify } from 'packages/uni-shared/src/vdom/Node'
import {
  ACTION_TYPE_DICT,
  type DictAction,
  type Dictionary,
} from '@dcloudio/uni-app-plus/constants'
import {
  createGetDict,
  decodeNodeJson,
} from '@dcloudio/uni-app-plus/view/framework/dom/decodeActions'
import {
  $,
  createElement,
  initPageScroll,
  onPageCreate,
  onPageCreated,
  onPageReady,
} from './page'
import { flushPostActionJobs } from '@dcloudio/uni-app-plus/view/framework/dom/scheduler'

export function onVdSync(actions: (PageAction | DictAction)[]) {
  if (__DEV__) {
    console.log(formatLog('onVdSync', actions))
  }
  const firstAction = actions[0]
  // page create
  if (firstAction[0] === ACTION_TYPE_PAGE_CREATE) {
    onPageCreateSync(firstAction)
  } else {
    onPageReady(() => onPageUpdateSync(actions))
  }
}

function onPageCreateSync(action: PageCreateAction) {
  return onPageCreate(action[1])
}

function onPageUpdateSync(actions: (PageAction | DictAction)[]) {
  const dictAction = actions[0]
  const getDict = createGetDict(
    dictAction[0] === ACTION_TYPE_DICT ? (dictAction[1] as Dictionary) : []
  )
  actions.forEach((action) => {
    switch (action[0]) {
      case ACTION_TYPE_PAGE_CREATE:
        return onPageCreate(action[1])
      case ACTION_TYPE_PAGE_CREATED:
        return onPageCreated()
      case ACTION_TYPE_CREATE:
        const parentNodeId = action[3]
        return createElement(
          action[1],
          getDict(action[2] as number),
          // 部分性能低的手机，createAction 与 insertAction 是分开的，导致根节点 parentNodeId 为 -1
          parentNodeId === -1 ? 0 : parentNodeId,
          action[4],
          decodeNodeJson(getDict, action[5] as UniNodeJSONMinify)
        )
      case ACTION_TYPE_INSERT:
        return $(action[1]).insert(
          action[2],
          action[3],
          decodeNodeJson(getDict, action[4] as UniNodeJSONMinify)
        )
      case ACTION_TYPE_REMOVE:
        return $(action[1]).remove()
      case ACTION_TYPE_SET_ATTRIBUTE:
        return $(action[1]).setAttr(
          getDict(action[2] as number),
          getDict(action[3] as number)
        )
      case ACTION_TYPE_REMOVE_ATTRIBUTE:
        return $(action[1]).removeAttr(getDict(action[2] as number))
      case ACTION_TYPE_ADD_EVENT:
        return $(action[1]).addEvent(getDict(action[2] as number), action[3])
      case ACTION_TYPE_ADD_WXS_EVENT:
        return $(action[1]).addWxsEvent(
          getDict(action[2] as number),
          getDict(action[3] as number),
          action[4]
        )
      case ACTION_TYPE_REMOVE_EVENT:
        return $(action[1]).removeEvent(getDict(action[2] as number))
      case ACTION_TYPE_SET_TEXT:
        return $(action[1]).setText(getDict(action[2] as number))
      case ACTION_TYPE_PAGE_SCROLL:
        return initPageScroll(action[1])
    }
  })
  flushPostActionJobs()
}
