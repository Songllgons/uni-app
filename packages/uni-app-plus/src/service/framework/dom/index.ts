import { ACTION_TYPE_EVENT, formatLog } from '@dcloudio/uni-shared'
import { getPageById } from '../page/getCurrentPages'
import { type EventAction, onNodeEvent } from './onNodeEvent'
import type UniPageNode from './Page'

export function onVdSync(actions: EventAction[], pageId: string) {
  // 从所有pages中获取
  const page = getPageById(parseInt(pageId))
  if (!page) {
    if (__DEV__) {
      console.error(formatLog('onVdSync', 'page', pageId, 'not found'))
    }
    return
  }
  const pageNode = (page as any).__page_container__ as UniPageNode
  actions.forEach((action) => {
    switch (action[0]) {
      case ACTION_TYPE_EVENT:
        onNodeEvent(action[1], action[2], pageNode)
        break
    }
  })
}
