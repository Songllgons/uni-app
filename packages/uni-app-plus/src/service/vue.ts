export { initApp } from '@dcloudio/uni-vue'
