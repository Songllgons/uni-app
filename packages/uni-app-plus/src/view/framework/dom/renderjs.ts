import {
  type ComponentOptions,
  type ComponentPublicInstance,
  createApp,
} from 'vue'
import { RENDERJS_MODULES, formatLog } from '@dcloudio/uni-shared'
import {
  createComponentDescriptorVm,
  getComponentDescriptor,
} from '@dcloudio/uni-core'

import type { UniNode } from './elements/UniNode'
import type { UniCustomElement } from './components'

export function initRenderjs(node: UniNode, moduleIds: Record<string, string>) {
  Object.keys(moduleIds).forEach((name) => {
    initRenderjsModule(node, moduleIds[name])
  })
}

export function destroyRenderjs(node: UniNode) {
  const { __renderjsInstances } = node.$ as UniCustomElement
  if (!__renderjsInstances) {
    return
  }
  Object.keys(__renderjsInstances).forEach((id) => {
    __renderjsInstances[id].$.appContext.app.unmount()
  })
}

function initRenderjsModule(node: UniNode, moduleId: string) {
  const options = getRenderjsModule(moduleId)
  if (!options) {
    return
  }
  const el = node.$ as UniCustomElement
  ;(el.__renderjsInstances || (el.__renderjsInstances = {}))[moduleId] =
    createRenderjsInstance(el, options)
}

function getRenderjsModule(moduleId: string) {
  const __renderjsModules =
    window[('__' + RENDERJS_MODULES) as '__renderjsModules']
  const module = __renderjsModules && __renderjsModules[moduleId]
  if (!module) {
    return console.error(formatLog('renderjs', moduleId + ' not found'))
  }
  return module as ComponentOptions | { default: ComponentOptions }
}

function createRenderjsInstance(
  el: UniCustomElement,
  options: ComponentOptions
): ComponentPublicInstance {
  options = options.default || options
  options.render = () => {}
  return createApp(options)
    .mixin({
      mounted() {
        this.$ownerInstance = getComponentDescriptor(
          createComponentDescriptorVm(el),
          false
        )
      },
    })
    .mount(document.createElement('div'))
}
