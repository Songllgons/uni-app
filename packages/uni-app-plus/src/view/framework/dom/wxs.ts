import { isArray, isFunction } from '@vue/shared'
import {
  WXS_MODULES,
  WXS_PROTOCOL,
  formatLog,
  getValueByDataPath,
} from '@dcloudio/uni-shared'
import {
  createComponentDescriptorVm,
  getComponentDescriptor,
} from '@dcloudio/uni-core'
import type { UniCustomElement } from './components'
import type { UniNode } from './elements/UniNode'

declare global {
  interface Window {
    __wxsModules: {
      [name: string]: Record<string, any>
    }
    __renderjsModules: {
      [name: string]: Record<string, any>
    }
  }
}

function getViewModule(moduleId: string, ownerEl: UniCustomElement) {
  const __wxsModules = window[('__' + WXS_MODULES) as '__wxsModules']
  const module = __wxsModules && __wxsModules[moduleId]
  if (module) {
    return module
  }
  if (ownerEl && ownerEl.__renderjsInstances) {
    return ownerEl.__renderjsInstances[moduleId]
  }
}

const WXS_PROTOCOL_LEN = WXS_PROTOCOL.length

export function invokeWxs(
  el: UniCustomElement | undefined,
  wxsStr: string,
  invokerArgs?: unknown[]
) {
  const [ownerId, moduleId, invoker, args] = parseWxs(wxsStr)
  const ownerEl = resolveOwnerEl(el!, ownerId)
  if (isArray(invokerArgs) || isArray(args)) {
    // methods
    const [moduleName, methodName] = invoker.split('.')
    return invokeWxsMethod(
      ownerEl,
      moduleId,
      moduleName,
      methodName,
      invokerArgs || args
    )
  }
  return getWxsProp(ownerEl, moduleId, invoker)
}

export function invokeWxsEvent(
  el: UniCustomElement,
  wxsStr: string,
  event: Record<string, any>
) {
  const [ownerId, moduleId, invoker] = parseWxs(wxsStr)
  const [moduleName, methodName] = invoker.split('.')
  const ownerEl = resolveOwnerEl(el, ownerId)
  return invokeWxsMethod(ownerEl, moduleId, moduleName, methodName, [
    wrapperWxsEvent(event, el),
    getComponentDescriptor(createComponentDescriptorVm(ownerEl), false),
  ])
}

function resolveOwnerEl(el: UniCustomElement, ownerId: number) {
  if (el.__ownerId === ownerId) {
    return el
  }
  let parentElement = el.parentElement as UniCustomElement
  while (parentElement) {
    if (parentElement.__ownerId === ownerId) {
      return parentElement
    }
    parentElement = parentElement.parentElement as UniCustomElement
  }
  return el
}

function parseWxs(wxsStr: string) {
  return JSON.parse(wxsStr.slice(WXS_PROTOCOL_LEN)) as [
    number,
    string,
    string,
    any
  ]
}

export function invokeWxsProps(
  wxsStr: string,
  el: UniCustomElement,
  newValue: unknown,
  oldValue: unknown
) {
  const [ownerId, moduleId, invoker] = parseWxs(wxsStr)
  const ownerEl = resolveOwnerEl(el, ownerId)
  const [moduleName, methodName] = invoker.split('.')
  return invokeWxsMethod(ownerEl, moduleId, moduleName, methodName, [
    newValue,
    oldValue,
    getComponentDescriptor(createComponentDescriptorVm(ownerEl), false),
    getComponentDescriptor(createComponentDescriptorVm(el), false),
  ])
}

function invokeWxsMethod(
  ownerEl: UniCustomElement,
  moduleId: string,
  moduleName: string,
  methodName: string,
  args: unknown[]
) {
  const module = getViewModule(moduleId, ownerEl)
  if (!module) {
    return console.error(
      formatLog('wxs', 'module ' + moduleName + ' not found')
    )
  }
  const method = (module as Record<string, any>)[methodName]
  if (!isFunction(method)) {
    return console.error(moduleName + '.' + methodName + ' is not a function')
  }
  return method.apply(module, args)
}

function getWxsProp(
  ownerEl: UniCustomElement,
  moduleId: string,
  dataPath: string
) {
  const module = getViewModule(moduleId, ownerEl)
  if (!module) {
    return console.error(formatLog('wxs', 'module ' + dataPath + ' not found'))
  }
  return getValueByDataPath(module, dataPath.slice(dataPath.indexOf('.') + 1))
}

export type WxsPropsInvoker = (newValue: unknown) => void

export function createWxsPropsInvoker(
  node: UniNode,
  wxsInvoker: string,
  value: unknown
): WxsPropsInvoker {
  let oldValue = value
  return (newValue: unknown) => {
    try {
      // 延迟到执行时再获取$，因为刚开始创建invoker时，组件的$还未准备好
      invokeWxsProps(wxsInvoker, node.$ as UniCustomElement, newValue, oldValue)
    } catch (e) {
      console.error(e)
    }
    oldValue = newValue
  }
}

function wrapperWxsEvent(event: Record<string, any>, el: UniCustomElement) {
  const vm = createComponentDescriptorVm(el)
  Object.defineProperty(event, 'instance', {
    get() {
      return getComponentDescriptor(vm, false)
    },
  })
  return event
}
