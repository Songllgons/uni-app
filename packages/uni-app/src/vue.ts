import type { Slots } from 'vue'

export function renderComponentSlot(
  slots: Slots,
  name: string,
  props: any | null = null
): any | null {
  if (slots[name]) {
    return slots[name]!(props)
  }
  return null
}
