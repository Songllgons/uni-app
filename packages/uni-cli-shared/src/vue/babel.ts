import type * as BabelCore from '@babel/core'
import type { PluginObj } from '@babel/core'
import { isBuiltInComponent } from '@dcloudio/uni-shared'

export function transformUniH5Jsx({ types }: typeof BabelCore): PluginObj {
  return {
    name: 'babel-plugin-uni-h5-jsx',
    visitor: {
      JSXOpeningElement({ node: { name } }) {
        if (types.isJSXIdentifier(name) && isBuiltInComponent(name.name)) {
          name.name = 'v-uni-' + name.name
        }
      },
    },
  }
}
