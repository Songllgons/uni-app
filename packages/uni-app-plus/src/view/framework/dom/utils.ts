import { JSON_PROTOCOL, WXS_PROTOCOL } from '@dcloudio/uni-shared'
import { isString } from '@vue/shared'
import type { UniCustomElement } from './components'
import type { UniComponent } from './components/UniComponent'
import { invokeWxs } from './wxs'

const JSON_PROTOCOL_LEN = JSON_PROTOCOL.length

export function decodeAttr(value: unknown, el?: UniCustomElement) {
  if (!isString(value)) {
    return value
  }
  if (value.indexOf(JSON_PROTOCOL) === 0) {
    value = JSON.parse(value.slice(JSON_PROTOCOL_LEN))
  } else if (value.indexOf(WXS_PROTOCOL) === 0) {
    value = invokeWxs(el, value)
  }
  return value
}

export function isCssVar(name: string) {
  return name.indexOf('--') === 0
}

export function isUniComponent(
  el: UniCustomElement | UniComponent
): el is UniComponent {
  // 不使用 instanceof 判断，因为需要引入 UniComponent，两者循环依赖
  return !!(el as UniComponent).addWxsEvent
}
