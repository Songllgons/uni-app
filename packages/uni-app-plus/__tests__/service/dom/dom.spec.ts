import { nextTick } from 'vue'
import {
  ACTION_TYPE_ADD_EVENT,
  ACTION_TYPE_CREATE,
  ACTION_TYPE_REMOVE,
  ACTION_TYPE_REMOVE_ATTRIBUTE,
  ACTION_TYPE_REMOVE_EVENT,
  ACTION_TYPE_SET_ATTRIBUTE,
  ACTION_TYPE_SET_TEXT,
  type CreateAction,
  EventModifierFlags,
  type SetAttributeAction,
  type UniEventListener,
} from '@dcloudio/uni-shared'
import type UniPageNode from '../../../src/service/framework/dom/Page'
import { createPageNode } from '../../../src/service/framework/dom/Page'
import {
  createElement,
  createTextNode,
  withModifiers,
} from '../../../../uni-app-vue/lib/service.runtime.esm'
import {
  ACTION_TYPE_DICT,
  type DictAction,
  setActionMinify,
} from '../../../src/constants'
import { decodeActions } from '../../../src/view/framework/dom/decodeActions'

describe('dom', () => {
  const pageId = 1
  const root = createPageNode(pageId, {
    css: true,
    route: 'pages/index/index',
    version: 1,
    locale: 'zh_CN',
    platform: 'ios',
    pixelRatio: 1,
    windowWidth: 375,
    disableScroll: false,
    onPageScroll: false,
    onPageReachBottom: false,
    onReachBottomDistance: 50,
    statusbarHeight: 24,
    windowTop: 0,
    windowBottom: 0,
  })
  test('minify = false', () => {
    setActionMinify(false)
    const viewElem = createElement('view', { pageNode: root })
    viewElem.setAttribute('id', 'view')
    root.appendChild(viewElem)
    viewElem.setAttribute('hidden', true)
    const { updateActions } = root
    const createElementAction = updateActions[0] as CreateAction
    expect(createElementAction[0]).toBe(ACTION_TYPE_CREATE)
    expect(createElementAction[1]).toBe(1)
    expect(createElementAction[2]).toBe('VIEW')
    expect(createElementAction[3]).toBe(0)
    expect(createElementAction[4]).toBe(-1)
    expect((createElementAction[5]!.a as any).id).toBe('view')
    // const addElementAction = updateActions[1] as InsertAction
    // expect(addElementAction[0]).toBe(ACTION_TYPE_INSERT)
    // expect(addElementAction[1]).toBe(1) // nodeId
    // expect(addElementAction[2]).toBe(0) // parentNodeId
    // expect(addElementAction[3]).toBe(-1) // index

    const setAttributeAction = updateActions[1] as SetAttributeAction
    expect(setAttributeAction[0]).toBe(ACTION_TYPE_SET_ATTRIBUTE)
    expect(setAttributeAction[1]).toBe(1)
    expect(setAttributeAction[2]).toBe('hidden')
    expect(setAttributeAction[3]).toBe(true)

    root.updateActions.length = 0
    viewElem.removeAttribute('hidden')
    const {
      updateActions: [removeAttributeAction],
    } = root
    expect(removeAttributeAction[0]).toBe(ACTION_TYPE_REMOVE_ATTRIBUTE)
    expect(removeAttributeAction[1]).toBe(1)
    expect(removeAttributeAction[2]).toBe('hidden')

    root.updateActions.length = 0
    viewElem.textContent = 'text'
    const {
      updateActions: [setTextAction],
    } = root
    expect(setTextAction[0]).toBe(ACTION_TYPE_SET_TEXT)
    expect(setTextAction[1]).toBe(1)
    expect(setTextAction[2]).toBe('text')

    root.updateActions.length = 0
    root.removeChild(viewElem)
    const {
      updateActions: [removeChildAction],
    } = root
    expect(removeChildAction[0]).toBe(ACTION_TYPE_REMOVE)
    expect(removeChildAction[1]).toBe(1)

    root.updateActions.length = 0
    const textNode = createTextNode('hello', { pageNode: root })
    root.appendChild(textNode)
    const {
      updateActions: [addTextNodeAction],
    } = root
    expect(addTextNodeAction[0]).toBe(ACTION_TYPE_CREATE)
    expect(addTextNodeAction[1]).toBe(2)
    expect(addTextNodeAction[2]).toBe('#text')
    expect(addTextNodeAction[3]).toBe(0)
    expect(addTextNodeAction[4]).toBe(-1)

    root.updateActions.length = 0
    const clickFn = () => {}
    textNode.addEventListener('click', clickFn)
    const {
      updateActions: [addEventListenerAction],
    } = root
    expect(addEventListenerAction[0]).toBe(ACTION_TYPE_ADD_EVENT)
    expect(addEventListenerAction[1]).toBe(2)
    expect(addEventListenerAction[2]).toBe('onClick')
    expect(addEventListenerAction[3]).toBe(0)

    root.updateActions.length = 0
    textNode.removeEventListener('click', clickFn)
    const {
      updateActions: [removeEventListenerAction],
    } = root
    expect(removeEventListenerAction[0]).toBe(ACTION_TYPE_REMOVE_EVENT)
    expect(removeEventListenerAction[1]).toBe(2)
    expect(removeEventListenerAction[2]).toBe('onClick')

    root.updateActions.length = 0
    const clickFn1 = withModifiers(() => {}, [
      'stop',
      'prevent',
    ]) as unknown as UniEventListener
    textNode.addEventListener('click', clickFn1, { capture: true })
    const {
      updateActions: [addEventListenerAction1],
    } = root
    expect(addEventListenerAction1[0]).toBe(ACTION_TYPE_ADD_EVENT)
    expect(addEventListenerAction1[1]).toBe(2)
    expect(addEventListenerAction1[2]).toBe('onClickCapture')
    const flag = addEventListenerAction1[3] as number
    expect(flag & EventModifierFlags.stop).toBeTruthy()
    expect(flag & EventModifierFlags.prevent).toBeTruthy()
    expect(flag & EventModifierFlags.self).toBeFalsy()
  })
  test('minify = true', () => {
    setActionMinify(true)

    const viewElem = createElement('view', { pageNode: root })
    viewElem.setAttribute('id', 'view')
    root.appendChild(viewElem)
    viewElem.setAttribute('hidden', true)
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    viewElem.removeAttribute('hidden')
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    viewElem.textContent = 'text'
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    root.removeChild(viewElem)
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    const textNode = createTextNode('hello', { pageNode: root })
    root.appendChild(textNode)
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    const clickFn = () => {}
    textNode.addEventListener('click', clickFn)
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    textNode.removeEventListener('click', clickFn)
    expectActions(root)

    root.dicts.length = 0
    root.updateActions.length = 0
    const clickFn1 = withModifiers(() => {}, [
      'stop',
      'prevent',
    ]) as unknown as UniEventListener
    textNode.addEventListener('click', clickFn1, { capture: true })
    expectActions(root)
    const flag = root.updateActions[0][3] as number
    expect(flag & EventModifierFlags.stop).toBeTruthy()
    expect(flag & EventModifierFlags.prevent).toBeTruthy()
    expect(flag & EventModifierFlags.self).toBeFalsy()
  })
  test('restore', (done) => {
    root.childNodes = []
    root.clear()
    const viewElem = createElement('view', { pageNode: root })
    viewElem.setAttribute('id', 'view')
    root.appendChild(viewElem)
    viewElem.setAttribute('hidden', true)

    const textNode = createTextNode('hello', { pageNode: root })
    root.appendChild(textNode)
    const clickFn1 = withModifiers(() => {}, [
      'stop',
      'prevent',
    ]) as unknown as UniEventListener
    textNode.addEventListener('click', clickFn1, { capture: true })
    nextTick(() => {
      root.restore()
      nextTick(done)
    })
  })
})

function expectActions(root: UniPageNode) {
  const actions = [
    [ACTION_TYPE_DICT, root.dicts] as DictAction,
    ...root.updateActions,
  ]
  expect(JSON.stringify(actions)).toMatchSnapshot()
  expect(JSON.stringify(decodeActions(actions))).toMatchSnapshot()
}
