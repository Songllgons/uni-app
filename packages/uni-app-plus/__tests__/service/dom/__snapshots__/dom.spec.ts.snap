// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dom minify = true 1`] = `"[[0,["VIEW","id","view","hidden",true]],[3,3,0,0,-1,{"a":[[1,2]]}],[6,3,3,4]]"`;

exports[`dom minify = true 2`] = `"[[0,["VIEW","id","view","hidden",true]],["create",3,"VIEW",0,-1,{"a":{"id":"view"}}],["setAttr",3,"hidden",true]]"`;

exports[`dom minify = true 3`] = `"[[0,["hidden"]],[7,3,0]]"`;

exports[`dom minify = true 4`] = `"[[0,["hidden"]],["removeAttr",3,"hidden"]]"`;

exports[`dom minify = true 5`] = `"[[0,["text"]],[10,3,0]]"`;

exports[`dom minify = true 6`] = `"[[0,["text"]],["setText",3,"text"]]"`;

exports[`dom minify = true 7`] = `"[[0,[]],[5,3]]"`;

exports[`dom minify = true 8`] = `"[[0,[]],["remove",3]]"`;

exports[`dom minify = true 9`] = `"[[0,["#text","hello"]],[3,4,0,0,-1,{"t":1}]]"`;

exports[`dom minify = true 10`] = `"[[0,["#text","hello"]],["create",4,"#text",0,-1,{"t":"hello"}]]"`;

exports[`dom minify = true 11`] = `"[[0,["onClick"]],[8,4,0,0]]"`;

exports[`dom minify = true 12`] = `"[[0,["onClick"]],["addEvent",4,"onClick",0]]"`;

exports[`dom minify = true 13`] = `"[[0,["onClick"]],[9,4,0]]"`;

exports[`dom minify = true 14`] = `"[[0,["onClick"]],["removeEvent",4,"onClick"]]"`;

exports[`dom minify = true 15`] = `"[[0,["onClickCapture"]],[8,4,0,3]]"`;

exports[`dom minify = true 16`] = `"[[0,["onClickCapture"]],["addEvent",4,"onClickCapture",3]]"`;
